### DISCUSSION TITLE
Instancing a Control Outside of Main Window Visual Tree

### DISCUSSION THREAD

**Discussion by DJGosnell** [AvaloniaUI/Avalonia#15811] [Q&A] [answered]
Is there a manual way to instance a `UserControl`, measure it, arrange it and then use `DrawingContextHelper.RenderAsync` to render the visuals another canvas context without attaching the `UserControl` directly to a Window?  I can instance the `UserControl` just fine with content, but the `VisualChildren` is then empty.  I see Logical descendants, but no `VisualChildren`.  I believe that the visual children all need to be Measured and Arranged before I can use `DrawingContextHelper.RenderAsync` to render them to another canvas.

```csharp
var uc = new UserControl() {
   Content = new TextBox() {
      Text = "TEST"
   }
};

uc.InitializeComponent();
uc.ApplyTemplate();
uc.Measure(Size.Infinity);
uc.Arrange(new Rect(uc.DesiredSize));

// This is always empty
uc.GetVisualChildren();
```

What is the proper way to instance a control?

If there is no way to do this, then that is something I need to know.

**Comment by maxkatz6** [comment, marked as answer]
You should be able to create a simple Window without showing it, to be able to render content. It will work as an actual root for your control.

**Reply by DJGosnell** [reply]
That's a great idea.  Thank you.

**Reply by Upanc** [reply]
Not working for me,Bounds aways {0,0,0,0}
